### DISCUSSION TITLE
Cross compilation by default

### DISCUSSION THREAD

**Discussion by Krakonos** [mesonbuild/meson#10904] [Q&A] [answered]
Hi!

I'm looking into replacing CMake with Meson for an embedded project that always cross-compiles using arm-none-eabi toolchain. As there is a lot of platform specific code, there is little chance of it working on another platform. I currently set toolchain file in CMake by default, so I don't have to specify it each time. I prefer the project builds with a simple `meson setup builddir && meson compile -C builddir`, not passing crossfile each time (and wondering why it fails if it's not passed). I'm looking through the documentation and could not find it. What's the suggested way to achieve it if it's even possible? I could always revert to a shell script to run meson, but that's not preventing devs from running meson directly not noticing the script.

Any advice is appreciated!

**Comment by eli-schwartz** [comment, marked as answer]
If nothing else, you can do
```meson
if not meson.is_cross_build()
    error('must be invoked with "meson setup--cross-file=./toolchain.ini" or via meson-wrapper.sh')
endif
```

This *does* make it impossible for devs to overlook the script.

I don't know off the top of my head if this is otherwise possible, unless maybe you can specify a cross file in default_options? Haven't tried it.

**Reply by xclaesse** [reply]
> unless maybe you can specify a cross file in default_options

No, cross/native files are not options, just command line arguments. That should maybe be changed, but currently there is no other way than adding an `error()`.

**Reply by JonasVautherin** [reply]
What about checking an environment variable like `MESON_CROSS_FILE`? I have a similar use-case as @Krakonos: I would like to enable meson in https://github.com/dockcross/dockcross/.

**Comment by jpakkane** [comment]
One of the design principles of Meson has always been to be as "functional" as possible. That is, once all the various options and other state is set, the configure phase "just runs" and should always produce the same output. Because of this you can't, for example, change the values of options from inside the build files. Doing so would make the whole thing unstable because you would be able to do bad things liked `set_option('some_boolean', not get_option('some_boolean'))` which would produce different results every other run.

Whether something is a cross build or not (and more specifically what machine it is targetting) is one of the pieces of state that is set "from the outside". While it would be convenient for some, it would open a whole different can of worms for many other projects.

**Reply by xclaesse** [reply]
Something I did years ago but got lost is to make cross_file a global built-in option. That way you can set it from project(... default_options: ...). I think it would solve this issue while still respecting current meson design.

**Reply by jpakkane** [reply]
There would be a limitation, namely you should not be able to change it afterwards (just like `backend').

It would also need refactoring, because cross files are set before build files are even parsed (IIRC).
